parse.lisp

```diff
@@ -453,14 +453,91 @@
       (destructuring-bind (&optional (et '*) (length '*)) rest
         (array-ctype :either et (list length) env))))
 
+(defvar *parse-extended-types* nil
+  "When `t', `specifier-ctype' will parse extended types. Use
+  `extended-specifier-ctype' instead of using this variable directly.")
+
+(defvar *env* nil
+  "This variable is for `specifier-ctype' to indirectly pass its environment to
+  extended type parsers. Ordinary lambda lists do not allow &environment, but
+  `specifier-ctype' should take an environment. Many functions get around this
+  by using &optional, but this does not work for `define-extended-type' since
+  extended types might already have &optional and may have &rest and &key as
+  well. It is not necessary to use this variable directly because &environment
+  is allowed in `define-extended-type'.")
+
+(defun remove-environment (lambda-list)
+  "Return(0) a new lambda list like LAMBDA-LIST but without the &environment
+  parameter. Return(1) the name of the removed &environment parameter."
+  (let (environment)
+    (values
+     (loop for keys on lambda-list
+           for key = (first keys)
+           if (eq key '&environment) do
+             (setq environment (second keys)
+                   keys (rest keys))
+           else collect key)
+     environment)))
+
+(defmacro define-extended-type (name lambda-list &key (documentation "") simple extended)
+  "Define a type NAME that can be used as a type specifier and as a constructor
+  for a custom ctype. The :simple expander is used by programs that only work
+  with type specifiers like `specifier-ctype'. The :extended expander is used by
+  programs that can take advantage of ctype extensions like
+  `extended-specifier-ctype'.
+
+  SIMPLE is a list of forms that return a type specifier that might not
+  completely represent the custom type.
+
+  EXTENDED is a list of forms that return a ctype that completely represents the
+  custom type.
+
+  Both the SIMPILE and the EXTENDED forms share the parameters of LAMBDA-LIST.
+
+  LAMBDA-LIST is an ordinary lambda list that also allows &environment."
+  (assert simple nil "simple form is required")
+  (assert extended nil "extended form is required")
+  `(progn
+     (deftype ,name ,lambda-list
+       ,documentation
+       ,@simple)
+     (setf (get ',name 'extended-type-parser)
```

Comment:
I use the symbol property list. I hope that's okay.

Reply:
It means the environment is ignored, but I have some thoughts on that I'll put elewhere